lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp

```diff
@@ -0,0 +1,252 @@
+//===- FuncEmissionPatterns.cpp - Func Dialect Emission Patterns ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This implements the emission patterns for the func dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "FuncEmissionPatterns.h"
+#include "../EmissionPrinter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+
+using namespace mlir::func;
+using namespace circt;
+using namespace circt::ExportSystemC;
+
+//===----------------------------------------------------------------------===//
+// Operation emission patterns.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Emit a func.func function. Users of the function arguments request an
+/// expression to be inlined and we simply return the name of the argument.This
+/// name has to be passed to this emission pattern via an array of strings
+/// attribute called 'argNames' because the emitter cannot do any name uniquing
+/// as it just emits the IR statement by statement. However, relying on an
+/// attribute for the argument names also has the advantage that the names
+/// can be preserved during a lowering pipeline and upstream passes have more
+/// control on how the arguments should be named (e.g. when they create a
+/// function and have some context to assign better names).
```

Comment:
The `argNames` seems like a necessary evil. I don't see a good way around this without requiring changes to upstream `func` dialect. The intent here is that names will need to be legalized via a pass before the conversion is run, yes?

Reply:
Correct (assuming conversion refers to emission).